### ISSUE TITLE
A method to rotate cubemap needed or something

### ISSUE THREAD

**Issue by makc** [mrdoob/three.js#11103]
Recently I had to deal with the models exported in the coordinates different from 3js. To cut long story short, it was far easier to rotate the camera than to change the model + code that depends on particular model orientation. However, I now had no way to make the environment map follow the camera rotation. Quick SO check shows that there are maybe 3 similar questions unanswered.

So what I did, in the end, was
```
THREE.ShaderChunk["envmap_fragment"] = THREE.ShaderChunk["envmap_fragment"]
    .split("reflectVec = reflect( cameraToVertex, worldNormal )")
    .join("reflectVec = transformDirection( reflect( cameraToVertex, worldNormal ), viewMatrix )");
```

but this is too hacky. Therefore, do you think we can at some point have some legit way to rotate cubemap in the material?

**Comment by WestLangley** [comment]
Maybe. What is hacky? The code injection or the contents of the injection?

I do not expect what you have implemented is correct; you should likely transform the direction by the coordinate system transform, not by the camera view matrix. But that is a separate issue.

In any event, one could argue it is reasonable to implement such a feature for the coordinate system z-up. That is a quarter-turn. We could have a z-up flag, I guess. Is that your use case?

If the model coordinate system is left-handed, or a reflection, then this gets more complicated.

**Comment by makc** [comment]
> What is hacky? The code injection or the contents of the injection?

blanket patch of all the shaders in the app for the single glass material does not feel right.

> We could have a z-up flag, I guess. Is that your use case?

basically they export from sketchup with YZ swap checkbox unchecked.

> transform the direction by the coordinate system transform

I could not find it in the uniforms

**Comment by WestLangley** [comment]
>blanket patch of all the shaders in the app for the single glass material does not feel right.

You do not have to blanket patch. See https://github.com/mrdoob/three.js/issues/10764#issuecomment-278507889.

>basically they export from sketchup with YZ swap checkbox unchecked

That is a reflection and rotation, I expect.

>I could not find it in the uniforms

That is because it is not there.

**Comment by NNskelly** [comment]
I'd second this request.  I think I'm noticing some other issues with env mapping in general which I'll file independently once I have a minimal repro case, but when the env map is representative of e.g. a real world location detailed in the environment and/or skybox, it seems ridiculous to have to orient the entire scene to line up with the angle at which the environment/reflection map wants to display, rather than being able to adjust the angle of the env map to match the scene.

I have tried using the u/v offset parameter, and while that appears to work for e.g. skyspheres with a latLong texture, it does not appear to do anything for the same texture applied as an env map...

The best hack I've found so far is to chop up the source image in an image editor and stitch it back at a different rotation before even bringing it into THREE, but that only really works for lat-long images that need to be adjusted laterally, not for vertical adjustments, cubemaps (by anything but 90 degrees), etc.

**Comment by WestLangley** [comment]
@NNskelly Cubemaps are axis-aligned. Try this:
```
scene.add( camera );
scene.rotation.set( heading, pitch, 0, 'YXZ' );
```

**Comment by NNskelly** [comment]
Is that equivalent to baking the heading/pitch into the root transformation matrix of a JSON scene file?  What I'm actually working on is an export/import/view pipeline from another tool, so having programmatic steps that need to occur after I've set up the actual geometry, materials, etc. is awkward.
In another test case, it looks like the cubemap is always camera-axis-aligned anyway, not world-axis-aligned, which is not iirc how it's normally done in other engines.

**Comment by mrdoob** [comment]
> In another test case, it looks like the cubemap is always camera-axis-aligned anyway, not world-axis-aligned, which is not iirc how it's normally done in other engines.

Do you mean that other engines apply the scene matrix to the cubemap?

**Comment by NNskelly** [comment]
> Do you mean that other engines apply the scene matrix to the cubemap?
That _sounds_ potentially correct, although it has been years since I've had to build a transformation matrix stack from scratch.  
Philosophically, though, what needs to come out of the lookup is the cubemap sample at the world-angle achieved by reflecting the world-space camera->fragment vector across the fragment normal, with the cubemap fixed in space as it were (the angle at which it is fixed can then be adjusted with a fixed offset, per the original issue).  This will cause any side of an object to appear to reflect whatever side of the "environment" it is facing.  Right now, however, the whole environment reflection appears to be pegged to the camera, which off the top of my head, suggests some transform is being applied symmetrically to both the camera and, erroneously, the cubemap, or conversely, that once you're down at the camera-space fragment part of the pipeline and everything is _implicitly_ pegged to the camera, there is some additional transform that needs to be applied to "the cubemap," which likely means the final cubemap lookup vector.

**Comment by mrdoob** [comment]
I'm not sure applying the `scene.matrix` to the cubemap would solve @makc issue though...

**Comment by NNskelly** [comment]
> I'm not sure applying the scene.matrix to the cubemap would solve @makc issue though...

Agreed.  Going back to the top, it actually looks like @makc 's starting point was a proposed fix for the issue I'm seeing- would the correct read of that line be that it is breaking out the reflection vector off the surface normal and applying the view matrix to it before injecting it back into the pipeline?  I should find a place to splice that into the A-Frame boilerplate I'm currently running on just to confirm.
Using that sort of approach to make the cubemap "fixed" relative to the world axes rather than the camera axes would at least allow a stopgap of laying out the rest of the scene to align with the fixed env map, but ideally, yes, as requested, a way to preemptively specify a fixed rotation on the cubemap would complete the solution.

Even foregoing @makc 's fix, tho, exposing hooks to affect the orientation of the cubemap directly would allow a stopgap of manually adjusting the cubemap angle at runtime so individual users could correct for any perceived discrepancies between their expectation and what it's actually doing without needing to go as far as hacking the shaders.

Also, thanks @WestLangley , while I still don't have an actual codebase to try those lines in, the notion of just spinning the camera around a mapped object to see what happens did crystalize my editor tests/repro in the tangential discussion https://github.com/mrdoob/three.js/issues/11367#issuecomment-303159114 - as it turns out, the angle of the camera does not appear to affect the reflection map; any changes in the reflection a surface shows seem to be purely a function of normal relative to camera, not camera relative to world.

**Comment by makc** [comment]
@NNskelly if you want to reproduce my setup,
```javascript
			var camera = new THREE.PerspectiveCamera (45, 1, 0.1 * sphere.radius, 10 * sphere.radius);
			camera.position.copy (sphere.center);
			camera.position.z += 4 * sphere.radius;

			var holder = new THREE.Object3D;
			holder.rotation.x = Math.PI / 2;
// hack to make the camera more centered
holder.position.y += 0.8 * sphere.radius;
			holder.position.z = sphere.radius / 2;
			holder.add (camera);
			scene.add (holder);
```
here's the piece of code from that project

**Comment by NNskelly** [comment]
No dice getting @makc 's hack to work just yet (incidentally, that's all just a js _string splice_?  That IS a bit crunchy, but props on figuring it out!).  I've pulled a copy of the editor source and am basically just bolting script fragment onto the end after everything else is loaded.  Is there an additional step of e.g. pushing the shader script back down to OpenGL after updating it?

**Comment by makc** [comment]
I just did it before everything else. Also, like @WestLangley said, it only resembles the effect of what's really have to be done, so don't rely on it.

**Comment by pailhead** [comment]
Would it make any sense to transform the cube that's actually drawing the cubemap, then render that into another cubemap and use that instead?

**Comment by NNskelly** [comment]
That might be just crazy enough to get the effect down.  It would require a couple extra steps, though, I think.  The cubemap "cube" is generally just a convenient input data representation for what ends up being a position-independent 360x180 degree angular color lookup, not a physical thing that exists in the world.  But provided there is a way to capture a runtime cubemap sample (sometimes referred to as a reflection probe), it may be possible to put the desired cubemap as an unlit emissive texture on an inwards-facing cube or sphere (basically a mini-skybox) somewhere outside the main scene, rotate that as desired, sample from the center of it to get a new texture, and then use that texture as the env map.  I suspect that would end up mirroring the map, though, so you'd probably need to do it twice to get back an unmirrored map at the right angle.  Would not be a great solution for real-time, but potentially for still renders.

**Comment by pailhead** [comment]
I have this:
https://www.npmjs.com/package/three-skybox

It's basically the same thing as the threejs cubemap with some slight optimizations. Did it before three could handle this internally. But, what it may be useful for, is experimenting with this kind of stuff. It basically does what you described, an emissive texture on an inwards facing cube, but does so outside the renderer.

Anyway, i was thinking of doing this only once.

```
loader.load( cubeMap , (texture)=>{
  myRotatedCubemap = rotateCubemap( cubemap , matrix || horizontalAngle )
```

**Comment by pailhead** [comment]
Also, just documenting this for:

https://github.com/mrdoob/three.js/pull/10791

Rather than "hacking":

>but this is too hacky. Therefore...
```
THREE.ShaderChunk["envmap_fragment"] = THREE.ShaderChunk["envmap_fragment"]
    .split("reflectVec = reflect( cameraToVertex, worldNormal )")
    .join("reflectVec = transformDirection( reflect( cameraToVertex, worldNormal ), viewMatrix )");
```

You'd just use the "glsl injection" feature of three.js and do whatever you want to which ever material you want without hacking.

@mrdoob ?

**Comment by NNskelly** [comment]
So I finally got enough bandwidth around my day job to actually dive in and try to fix this.

The file \src\renderers\shaders\ShaderChunk\envmap_fragment.glsl has all the goods; looks like they get packed into strings in the line ~563x strata of the built three.js.

As to my own issue, it looks like I may have had a misunderstanding of what "Spherical" maps are even supposed to be.  I got through a quick rewrite of the `#elif defined( ENVMAP_TYPE_SPHERE )` block starting on line 37 of the glsl, only to notice that the rewrite looked suspiciously similar to the `#elif defined( ENVMAP_TYPE_EQUIREC )` block above it.

@mrdoob What are "spherical" maps intended for?  It looks like the lat-long maps referred to as "spherical" environment maps in e.g. Unity and Unreal are in fact most accurately supported by the Equirectangular mappings.
The documentation at https://threejs.org/docs/index.html#api/constants/Textures says simply 

> EquirectangularReflectionMapping and EquirectangularRefractionMapping are for use with an equirectangular environment map.
> SphericalReflectionMapping is for use with a spherical reflection map.

with the example links leading either nowhere or back to the same definitions page...

To @makc 's original issue, rotating an equirec map along the vertical ought to be as simple as adding a new term e.g.
```
#elif defined( ENVMAP_TYPE_EQUIREC )

		vec2 sampleUV;
		sampleUV.y = saturate( flipNormal * reflectVec.y * 0.5 + 0.5 );
		sampleUV.x = atan( flipNormal * reflectVec.z, flipNormal * reflectVec.x ) * RECIPROCAL_PI2 + 0.5 + rotation_factor_between_0_and_1;
		vec4 envColor = texture2D( envMap, sampleUV );
```
and making sure the texture sampling is set to wrap on u/x.

Rotating an arbitrary map lookup would probably be done most simply by tweaking the original reflectVec on or about the line makc first proposed (12 and/or 16 of the glsl).  After getting the initial reflectVec value, just pop a
```
reflectVec = shinyNewRotMat * reflectVec;
```

where shinyNewRotMat is the combined and inverted(?) result of the various rotations you want to apply to the map cube (per e.g. http://mathworld.wolfram.com/RotationMatrix.html )
If the desired env rotation is in fact the camera rotation, then you can probably just rip that matrix out of some earlier part of the stack rather than rebuilding it.  Using the final view matrix would get close if the camera is near the origin, but likely introduce some weird distortions from the translation and perspective frustum.  Wrapping the spliced in a normalize() might mitigate the latter to some extent, but yes, would not be strictly correct.

**Comment by pailhead** [comment]
There should be some `#ifdef CUBEMAP_ROTATION` controlling this since people may not want another matrix multiplication to be happening for every pixel?

**Comment by NNskelly** [comment]
very valid point, pailhead.  I would also imagine that a quaternion would be sufficient instead of a full 3x3 or 4x4 matrix input, although I'm still not 100% confident at building/using them correctly.

**Comment by manthrax** [comment]
Alternatively could just write a helper function to re-render a cubemap into a new cubemap with a supplied transformation. Granted it might be lossy, but saves on code complexity.

Could be added as a method of CubeMapCamera or similar.

**Comment by Mugen87** [comment]
Closing in favor of #16328.
